File: ios/Views/YVPBibleReaderView.swift

import ExpoModulesCore
import YouVersionPlatform
import SwiftUI

class BibleReaderViewProps: ExpoSwiftUI.ViewProps {
    // Bible reference
    @Field var versionId: Int
    @Field var bookUSFM: String
    @Field var chapter: Int
    @Field var verse: Int?
    @Field var verseStart: Int?
    @Field var verseEnd: Int?
    
    @Field var appName: String
    @Field var signInMessage: String
    @Field var hasReference: Bool
}

struct YVPBibleReaderView: ExpoSwiftUI.View, ExpoSwiftUI.WithHostingView {
    @ObservedObject var props: BibleReaderViewProps
    
    init(props: BibleReaderViewProps) {
        self.props = props
    }
    
    var body: some View {
        BibleReaderView(
            reference: bibleReference(),
            appName: props.appName,
            signInMessage: props.signInMessage
        )
    }
    
    func bibleReference() -> BibleReference? {
        if let start = props.verseStart, let end = props.verseEnd {
            return BibleReference(
                versionId: props.versionId,
                bookUSFM: props.bookUSFM,
                chapter: props.chapter,
                verseStart: start,
                verseEnd: end
            )
        }
        
        if props.hasReference {
            return BibleReference(
                versionId: props.versionId,
                bookUSFM: props.bookUSFM,
                chapter: props.chapter,
                verse: props.verse
            )
        }
        
        return nil
    }
}

File: android/src/main/java/com/youversion/reactnativesdk/views/YVPBibleReaderView.kt
package com.youversion.reactnativesdk.views

import androidx.compose.runtime.Composable
import com.youversion.platform.core.bibles.domain.BibleReference
import com.youversion.platform.reader.BibleReader
import expo.modules.kotlin.views.ComposeProps

const val DEFAULT_BEREAN_STANDARD_BIBLE_VERSION = 3034

data class BibleReaderViewProps(
    val versionId: Int? = null,
    val bookUSFM: String? = null,
    val chapter: Int? = null,
    val verse: Int? = null,
    val verseStart: Int? = null,
    val verseEnd: Int? = null,
    val appName: String? = null,
    val signInMessage: String? = null,
    val hasReference: Boolean? = null
) : ComposeProps

@Composable
fun YVPBibleReaderView(props: BibleReaderViewProps) {
    BibleReader(
        appName = props.appName ?: "",
        appSignInMessage = props.signInMessage ?: "",
        bibleReference = bibleReference(props),
    )
}

fun bibleReference(props: BibleReaderViewProps): BibleReference? {
    val start = props.verseStart
    val end = props.verseEnd

    if (start != null && end != null) {
        return BibleReference(
            versionId = props.versionId ?: DEFAULT_BEREAN_STANDARD_BIBLE_VERSION,
            bookUSFM = props.bookUSFM ?: "JHN",
            chapter = props.chapter ?: 1,
            verseStart = start,
            verseEnd = end
        )
    }

    if (props.hasReference == true) {
        return BibleReference(
            versionId = props.versionId ?: DEFAULT_BEREAN_STANDARD_BIBLE_VERSION,
            bookUSFM = props.bookUSFM ?: "JHN",
            chapter = props.chapter ?: 1,
            verse = props.verse
        )
    }

    return null
}
